src/lib/frontend/frontend.ml

```diff
@@ -250,29 +258,105 @@ module Make(SAT : Sat_solver_sig.S) : S with type sat_env = SAT.t = struct
       let status =
         if timeout != NoTimeout then (Timeout (Some d))
         else (Unknown (d, env))
       in
       print_status status (Steps.get_steps ());
-      print_model env (Some timeout);
       if timeout != NoTimeout then raise Util.Timeout;
-      env , consistent, dep
+      process_unknown
+        print_status used_context consistent_dep_stack
+        env dep d (Some timeout)
 
     | Util.Timeout as e ->
       print_status (Timeout (Some d)) (Steps.get_steps ());
-      print_model env None;
+      (* dont call 'process_unknown' in this case. Timeout stops
+         all-models listing *)
+      print_model (SAT.get_model env) None;
       raise e
 
+  and process_unknown
+      print_status used_context consistent_dep_stack env dep d timeout_kind =
+    match d.st_decl with
+    | Assume _ | PredDef _ | RwtDef _ | Push _ | Pop _ | ThAssume _->
+      (* cannot raise Sat or Unknown in this case *)
+      assert false
+
+    | Query(n, _, sort) ->
+      (* 1. check if we are in all-sat mode, and build the filter
+         that'll be applied on the boolean model *)
+      let all_sat_mode =
+        match sort with
+        | AllSat l ->
+          (* 1.A SMT command check-all-sat *)
+          Some (List.fold_left
+                  (fun acc s ->
+                     (*transform string to propositional vars (E.t)*)
+                     let t = E.mk_term (Symbols.name s) [] Ty.Tbool in
+                     E.Set.add t acc
+                  )E.Set.empty l)
+
+        | Thm | Sat ->
+          (* 1.B if user rather set option --alll-models: empty
+             filter. Otherwise, all_sat_mode = None *)
+          if Options.get_all_models () then Some E.Set.empty
+          else None
+
+        | Cut | Check ->
+          (* 1.3 all_sat_mode = None for cut and check *)
+          None
+      in
+      let m = match SAT.get_model env with
+        | None ->
+          if all_sat_mode != None then
+            (* all-sat enabled but interpretation disabled. No timeout
+               here! Return a model with just the propositional part
+            *)
+            Some (lazy {
+                Models.propositional = SAT.get_propositional_model env;
+                constants = Profile.P.empty;
+                functions = Profile.P.empty;
+                arrays = Profile.P.empty;
+                objectives = Util.MI.empty }
+              )
+          else
+            None
+        | Some _ as md -> md
+      in
+      match m, all_sat_mode with
+      | Some m, Some filter ->
+        (* 1. case where all-bool-models wrt. given filter is
+           requested *)
+        let m = Lazy.force m in
+        let propositional = filter_by_all_sat m.propositional filter in
+        let m = { m with propositional } in
+        print_model (Some (lazy m)) timeout_kind;
+        let f =
+          Expr.Set.fold
+            (fun e acc -> E.mk_and e acc false 0) propositional E.vrai
+        in
```

Comment:
Add some explanation on what is done here. `not f` (propositional model) is added as a query to force resolution `-` the previous sat model